packages/pyright-internal/src/analyzer/importResolver.ts

```diff
@@ -87,6 +85,7 @@ export class ImportResolver {
         if (this.fileSystem instanceof PyrightFileSystem) {
             this.fileSystem.clearPartialStubs();
         }
+        this.fileSystem.invalidateCache();
```

Comment:
I assume any change from file watcher (inside of lib folder + workspace root + extra path) will cause invalidate cache to be called?

Reply:
assuming it will get invalidated when the file watcher said something is changed, when would we ever want cache == false?

Reply:
The caller is responsible for calling invalidate. In this case, the import resolver is the caller, and it has enough knowledge to invalidate at appropriate times. Most other consumers of the vfs interface do not. Plus, there are cases where it's not safe to use caching, like in our file-based cancellation.

Reply:
I initially thought about doing the caching entirely within the import resolver, but there were a few related modules (like pyTypedUtils) that also needed to participate. Maybe it would still be better to do the caching in the import resolver. That would address most of the concerns you and Jake are making.

Reply:
Now that you describe this as "the caller calls invalidate", I understand your implementation a lot better. A user of the FS can audit all of its calls, then call invalidate when it thinks something's wrong. The caches may be getting invalidated more often than needed if everyone does it, but it at least leads to a slow move of things to be cache-aware and then we can make some other change to improve it.

Reply:
But I do see the inconsistent cache system as a potential bump in the road, yeah...

Reply:
I think I'll code a second version that includes all the caching in the import resolver. We can then compare/contrast the two approaches. I think you're both making good points, and the alternative approach would address all of them.

Reply:
Another motivator for making the import resolver cache something is that I think the import resolver is the only code that has to rescan the same things over and over and over again; the rest of the code either reads files of known paths, or is the indexer, which does a one-pass over a directory recursively until it's requested to do it again. The fact that we got this giant perf boost just by opting only the import resolver in at least matches that model.

(But, having the import resolver do a bunch of work may not be super fun either; we'll see, I guess...)

Reply:
I did a similar thing in indexer (https://github.com/microsoft/pyrx/blob/main/packages/pylance-internal/src/packageScanner.ts#L312)

I got a big perf win from it, so I actually tried it on import resolver before (got similar perf win) but dropped it since I wasn't sure when to invalidate the cache (indexer was easy since it runs once and thrown away)

Reply:
did a similar thing in indexer (packageScanner.ts#L312)

I got a big perf win from it, so I actually tried it on import resolver before (got similar perf win) but dropped it since I wasn't sure when to invalidate the cache (indexer was easy since it runs once and thrown away)